src/transformers/models/bark/modeling_bark.py

```diff
@@ -0,0 +1,1429 @@
+        beam_idx at every generation step.
+        """
+        # Necessary for beam_search
+        return tuple(
+            tuple(past_state.index_select(0, beam_idx.to(past_state.device)) for past_state in layer_past)
+            for layer_past in past_key_values
+        )
+
+
+@add_start_docstrings(
+    "Bark sub-module at the core of the semantic sub-model. It shares the same architecture than the coarse model. It is a GPT-2 like autoregressive model with a language modeling head on top.",
+    BARK_MODULE_START_DOCSTRING,
+)
+class BarkSemanticModule(BarkCausalModule):
+    base_model_prefix = "semantic"
+    config_class = BarkSemanticConfig
+
+
+@add_start_docstrings(
+    "Bark sub-module at the core of the coarse acoustics sub-model. It shares the same architecture than the semantic model. It is a GPT-2 like autoregressive model with a language modeling head on top.",
+    BARK_MODULE_START_DOCSTRING,
+)
+class BarkCoarseAcousticsModule(BarkCausalModule):
+    base_model_prefix = "coarse_acoustics"
+    config_class = BarkCoarseAcousticsConfig
+
+
+@add_start_docstrings(
+    "Bark sub-module at the core of the fine acoustics sub-model. It is a non-causal GPT-like model with 8 embedding layers and language modeling heads, one for each codebook.",
+    BARK_MODULE_START_DOCSTRING,
+)
+class BarkFineAcousticsModule(BarkModulePreTrainedModel):
+    base_model_prefix = "fine_acoustics"
+    config_class = BarkFineAcousticsConfig
+    main_input_name = "codebook_idx"
+    _tied_weights_keys = []
+    _keys_to_ignore_on_load_missing = []
+
+    def __init__(self, config):
+        # non-causal gpt-like model with one embedding layer and one lm_head for each codebook of Encodec
+        super().__init__(config)
+        self.config = config
+
+        # initialize a modified non causal GPT-like model
+        # note that for there is one embedding layer and one lm_head for each codebook of Encodec
+        self.transformer = nn.ModuleDict(
+            {
+                "wtes": nn.ModuleList(
+                    [nn.Embedding(config.input_vocab_size, config.hidden_size) for _ in range(config.n_codes_total)]
+                ),
+                "wpe": nn.Embedding(config.block_size, config.hidden_size),
+                "drop": nn.Dropout(config.dropout),
+                "h": nn.ModuleList([BarkBlock(config, is_causal=False) for _ in range(config.num_layers)]),
+                "ln_f": nn.LayerNorm(config.hidden_size),
+            }
+        )
+        self.lm_heads = nn.ModuleList(
+            [
+                nn.Linear(config.hidden_size, config.output_vocab_size, bias=False)
+                for _ in range(config.n_codes_given, config.n_codes_total)
+            ]
+        )
+        self.gradient_checkpointing = False
+        self.n_codes_total = config.n_codes_total
+
+        # Initialize weights and apply final processing
+        self.post_init()
```

Comment:
This is going to call `tie_weights()` for us if we're initialising the weights. If we're loading from pre-trained, then the weights will be tied anyway, so no need to explicitly tie the weights below